### PR TITLE
[python] corrects ploting of empty diagrams + plot instead of scatter

### PR DESCRIPTION
**Pull request by hschreiber**
Fixes Issue #1264

Example of code which was not working (because dim 2 is empty):

```python
import numpy as np
from gudhi.sklearn.rips_persistence import RipsPersistence
from gudhi import plot_persistence_diagram

rips_transformer = RipsPersistence(
        homology_dimensions=range(3),
        input_type="point cloud",
        homology_coeff_field=2
    )
pc = np.array([[0,2], [1,5], [4,0.5], [4,5]])
rips_transformer.fit(pc)
rips_target = rips_transformer.transform([pc])
plot_persistence_diagram(rips_target[0])
```

**Edit:** I also replaced the scatter plot with a plot plot to make the plotting faster (tested with a Rips filtration of 500 points). In a very naïve way (I didn't wanted to spend too much time on it and it works already faster this way).

### CODE REVIEW

File: src/python/gudhi/persistence_graphical_tools.py [hunk 1]
```diff
@@ -79,8 +80,9 @@ def _format_handler(a):
     # Iterable of array
     try:
         for elt in a:  # check that death values have correct type
-            if not isinstance(elt[0][1], (np.floating, float, np.integer, int)):
-                raise TypeError("Should be a list of (birth, death)")
+            if len(elt) != 0:
+                if not isinstance(elt[0][1], (np.floating, float, np.integer, int)):
```
Comment (mglisse) [inline src/python/gudhi/persistence_graphical_tools.py]: I think numpy recommends using `np.issubdtype` instead of isinstance (and then as a bonus `np.floating` magically handles `float`).

File: src/python/gudhi/persistence_graphical_tools.py [hunk 2]
```diff
@@ -89,6 +91,30 @@ def _format_handler(a):
     return a, 0
 
 
+# only necessary because _format_handler does not directly decompose everything into xd, yd arrays for plotting
```
Comment (mglisse) [inline src/python/gudhi/persistence_graphical_tools.py:94]: Maybe it should? I think @VincentRouvreau added a conversion from the new type to the old type because that was the easiest way to add support for the new type, but long term it should be the reverse.

File: src/python/gudhi/persistence_graphical_tools.py [hunk 3]
```diff
@@ -369,9 +396,25 @@ def plot_persistence_diagram(
...
     y = [death if death != float("inf") else infinity for (dim, (birth, death)) in persistence]
-    c = [colormap[dim] for (dim, (birth, death)) in persistence]
 
-    axes.scatter(x, y, alpha=alpha, color=c)
+    i = 0
+    for d in range(len(sizes)):
+        if sizes[d] != 0:
+            j = i + sizes[d]
+            xd = x[i:j]
+            yd = y[i:j]
```
Comment (mglisse) [inline src/python/gudhi/persistence_graphical_tools.py:416]: Does something sort the points by dimension before this?

File: src/python/gudhi/persistence_graphical_tools.py (hunk 2 above)
Comment (hschreiber) [reply src/python/gudhi/persistence_graphical_tools.py:94]: `format_handler_` is used by `plot_persistence_barcode`, `plot_persistence_diagram` and `plot_persistence_density` and they all use the output a little bit differently, so I see why this was the simplest way to adapt the possible barcode formats. But it feels like a lot of useless copying/moving around, i.e. not very optimal for all three. On the long run, it could be better to have a specialized version for all three, even if it is more annoying to maintain. And then, methods/loops like this one would become unnecessary.

File: src/python/gudhi/persistence_graphical_tools.py (hunk 3 above)
Comment (hschreiber) [reply src/python/gudhi/persistence_graphical_tools.py:416]: Oh, good point, no, there isn't for the input format `(dim, (b, d))`. I have to correct this.

Comment (hschreiber) [reply src/python/gudhi/persistence_graphical_tools.py:416]: Done, but I would redo some benchmarking to be sure that I did not added to much overhead with the sort (and the sizes) compared to the old version without all this.

Comment (hschreiber) [reply src/python/gudhi/persistence_graphical_tools.py:416]: Ok, the times are still right:

Rips 1000 random points, 1598 bars

#### array or array of (b,d)

before: 28.1 ms ± 833 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
after: 19.3 ms ± 492 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)

#### array of (b,d)

before: 26.7 ms ± 457 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
after: 17.6 ms ± 35.8 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

#### (dim, (b,d))

before: 23 ms ± 557 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
after: 15.1 ms ± 182 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

File: src/python/gudhi/persistence_graphical_tools.py (hunk 1 above)
Comment (VincentRouvreau) [reply src/python/gudhi/persistence_graphical_tools.py]:
```suggestion
                elt_type = type(elt[0][1])
                if not (np.issubdtype(elt_type, np.floating) or np.issubdtype(elt_type, np.integer)):
```
seems ok to me

File: src/python/gudhi/persistence_graphical_tools.py (hunk 2 above)
Comment (VincentRouvreau) [reply src/python/gudhi/persistence_graphical_tools.py:94]: Yes, that's not for the scope of this PR, but this is a reminder of #395 (and maybe #354) for the new persistence format

File: src/python/gudhi/persistence_graphical_tools.py [hunk 4]
```diff
@@ -367,11 +394,29 @@ def plot_persistence_diagram(
     # line display of equation : birth = death
     axes.plot([axis_start, axis_end], [axis_start, axis_end], linewidth=1.0, color="k")
 
+    if input_type == 0:
+        sorted(persistence, key=lambda x: x[0])
```
Comment (mglisse) [inline src/python/gudhi/persistence_graphical_tools.py]: Did you test that?

Comment (hschreiber) [reply src/python/gudhi/persistence_graphical_tools.py]: Well, I tested it outside of the method but not within, which was an error. I really need to get used to Pythons way of dealing with inputs.
Anyway, it should be corrected now.

I also added tests for it, even though they would not detect the previous error. I just thought it could be useful if we do bigger changes at some point.
